logbus/logstreamer/logstreamer.go

```diff
@@ -41,88 +30,40 @@ type LogBus interface {
 	RemoveSubscriber(logbus.Subscriber)
 }
 
-// Opt is an option function, used to adjust optional attributes of a
-// LogStreamer during New().
-type Opt func(*LogStreamer) *LogStreamer
-
-// WithBuffer ensures that there is a buffer for at least size messages on the
-// cloud.Deltas value that the LogStreamer sends to StreamLogs.
-func WithBuffer(size int) Opt {
-	return func(l *LogStreamer) *LogStreamer {
-		l.deltas.chSize = size
-		return l
-	}
-}
-
-// LogStreamer is a log streamer. It uses the cloud client to send
+// LogStreamer is a log streamer. It uses the cloud client to Write
 // log deltas to the cloud. It retries on transient errors.
 type LogStreamer struct {
-	bus     LogBus
 	c       CloudClient
 	buildID string
-	doneCh  chan struct{}
 	errors  []error
 
 	mu        sync.Mutex
 	cancelled bool
-	deltas    deltasIter
+	deltas    *deltasIter
 }
 
 // New creates a new LogStreamer.
-func New(ctx context.Context, bus LogBus, c CloudClient, initialManifest *logstream.RunManifest, opts ...Opt) *LogStreamer {
+func New(c CloudClient, buildID string, deltas *deltasIter) *LogStreamer {
 	ls := &LogStreamer{
-		bus:     bus,
 		c:       c,
-		buildID: initialManifest.GetBuildId(),
-		doneCh:  make(chan struct{}),
-		deltas: deltasIter{
-			chSize: DefaultBufferSize,
-			init:   initialManifest,
-		},
-	}
-	for _, o := range opts {
-		ls = o(ls)
+		buildID: buildID,
+		deltas:  deltas,
 	}
-	go ls.retryLoop(ctx)
 	return ls
 }
 
-func (ls *LogStreamer) retryLoop(ctx context.Context) {
-	defer close(ls.doneCh)
-	const maxRetry = 10
-	for i := 0; i < maxRetry; i++ {
-		retry, err := ls.tryStream(ctx)
-		if err == nil {
-			return
-		}
-		if i == maxRetry-1 {
-			retry = false
-		}
-		if !retry {
-			ls.errors = append(ls.errors, err)
-			return
-		}
-		fmt.Fprintf(os.Stderr, "transient error streaming logs: %v\n", err)
-	}
-}
-
-func (ls *LogStreamer) tryStream(ctx context.Context) (bool, error) {
+// Stream will attempt to stream deltas from the deltasIter to the Cloud
+// This also returns a bool indicating whether the error that occurred should be considered retryable
+func (ls *LogStreamer) Stream(ctx context.Context) (bool, error) {
 	ctxTry, cancelTry := context.WithCancel(ctx)
 	defer cancelTry()
 	ls.mu.Lock()
 	if ls.cancelled {
-		// TODO (vladaionescu): It would be nice if on cancellation we could
-		// 						still go through the entire retry loop.
-		//						This would require that we close ls.ch on each
-		//						retry somehow safely.
```

Comment:
I'm pretty sure the new wait introduced from the iterator `Done` handles this.
We also seem to stream double the formatted logs to the iterator than we need.